exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/TypeInferenceUtils.java

```diff
@@ -0,0 +1,571 @@
+    final SqlTypeName sqlTypeName = relDataType.getSqlTypeName();
+    TypeProtos.MinorType minorType = CALCITE_TO_DRILL_MAPPING.get(sqlTypeName);
+    if(minorType == null) {
+      minorType = TypeProtos.MinorType.LATE;
+    }
+    return minorType;
+  }
+
+  /**
+   * Give the name and DrillFuncHolder list, return the inference mechanism.
+   */
+  public static SqlReturnTypeInference getDrillSqlReturnTypeInference(
+      final String name,
+      final List<DrillFuncHolder> functions) {
+    switch(name.toUpperCase()) {
+      case "DATE_PART":
+        return DrillDatePartSqlReturnTypeInference.INSTANCE;
+
+      case "SUM":
+        return new DrillSumSqlReturnTypeInference(functions);
+
+      case "COUNT":
+        return DrillCountSqlReturnTypeInference.INSTANCE;
+
+      case "CONCAT":
+        return DrillConcatSqlReturnTypeInference.INSTANCE;
+
+      case "LENGTH":
+        return DrillLengthSqlReturnTypeInference.INSTANCE;
+
+      case "LPAD":
+      case "RPAD":
+      case "LTRIM":
+      case "RTRIM":
+      case "BTRIM":
+        return DrillPadTrimSqlReturnTypeInference.INSTANCE;
+
+      case "CONVERT_TO":
+        return DrillConvertToSqlReturnTypeInference.INSTANCE;
+
+      case "EXTRACT":
+        return DrillExtractSqlReturnTypeInference.INSTANCE;
+
+      case "CAST":
+        return DrillCastSqlReturnTypeInference.INSTANCE;
+
+      case "FLATTEN":
+      case "KVGEN":
+      case "CONVERT_FROM":
+        return DrillDeferToExecSqlReturnTypeInference.INSTANCE;
+
+      default:
+        return new DrillDefaultSqlReturnTypeInference(functions);
+    }
+  }
+
+  private static class DrillDefaultSqlReturnTypeInference implements SqlReturnTypeInference {
+    private final List<DrillFuncHolder> functions;
+
+    public DrillDefaultSqlReturnTypeInference(List<DrillFuncHolder> functions) {
+      this.functions = functions;
+    }
+
+    @Override
+    public RelDataType inferReturnType(SqlOperatorBinding opBinding) {
+      final RelDataTypeFactory factory = opBinding.getTypeFactory();
+      if (functions.isEmpty()) {
+        return factory.createTypeWithNullability(
+            factory.createSqlType(SqlTypeName.ANY),
+            true);
+      }
+
+      boolean allBooleanOutput = true;
+      for (DrillFuncHolder function : functions) {
+        if (function.getReturnType().getMinorType() != TypeProtos.MinorType.BIT) {
+          allBooleanOutput = false;
+          break;
+        }
+      }
+      if (allBooleanOutput) {
+        return factory
+            .createSqlType(SqlTypeName.BOOLEAN);
+      }
+
+      // The following logic is just a safe play:
+      // Even if any of the input arguments has ANY type,
+      // it "might" still be possible to determine the return type based on other non-ANY types
+      for (RelDataType type : opBinding.collectOperandTypes()) {
+        if (type.getSqlTypeName() == SqlTypeName.ANY || type.getSqlTypeName() == SqlTypeName.DECIMAL) {
```

Comment:
Can you explain why you will return ANY, if one of the input type is DECIMAL? Probably put some comments here.